core/src/main/java/org/hisp/dhis/android/core/arch/api/fields/internal/Fields.kt

```diff
@@ -25,58 +25,50 @@
  *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
+package org.hisp.dhis.android.core.arch.api.fields.internal
 
-package org.hisp.dhis.android.core.arch.api.fields.internal;
+internal data class Fields<T>(val fields: List<Property<T, *>>) {
 
-import com.google.auto.value.AutoValue;
+    class Builder<T> internal constructor() {
+        val fields: MutableList<Property<T, *>> = mutableListOf()
 
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
-
-import androidx.annotation.NonNull;
-
-@AutoValue
-public abstract class Fields<T> {
-
-    @NonNull
-    public abstract List<Property<T, ?>> fields();
-
-    @NonNull
-    public static <K> Fields.Builder<K> builder() {
-        return new Builder<>();
-    }
-
-    public static class Builder<T> {
-        private final List<Property<T, ?>> fields;
-
-        Builder() {
-            this.fields = new ArrayList<>();
+        @SafeVarargs
+        fun fields(vararg properties: Property<T, *>): Builder<T> {
+            require(properties.isNotEmpty()) { "properties should not be empty" }
+            fields.addAll(properties)
+            return this
         }
 
-        @SafeVarargs
-        public final Builder<T> fields(@NonNull Property<T, ?>... properties) {
-            if (properties == null || properties.length == 0) {
-                throw new IllegalArgumentException("properties == null or properties.length == 0");
-            }
+        fun fields(properties: Collection<Property<T, *>>): Builder<T> {
+            require(properties.isNotEmpty()) { "properties should not be empty" }
+            fields.addAll(properties)
+            return this
+        }
 
-            fields.addAll(Arrays.asList(properties));
-            return this;
+        fun build(): Fields<T> {
+            return Fields(fields.toList())
         }
+    }
 
-        public final Builder<T> fields(@NonNull Collection<Property<T, ?>> properties) {
-            if (properties == null || properties.isEmpty()) {
-                throw new IllegalArgumentException("properties null or empty collection");
+    fun generateString(properties: List<Property<*, *>>?): String {
```

Comment:
It is a bit weird to have to call it as `.generateString(null)` from the outside. The parameter is only needed for the recursion, isn't it? 

I can see two possible approaches to avoid it:
- Make the parameter optional, such as `generateString(properties: List<Property<*,*>>? = null)`. In this way, it might be called as `.generateString()`.
- Create a private method in the companion object with this logic, maybe called `private fun generateStringFromFields()` and then call it from here as `generateStringFromFields(fields)`.